src/gui/guiFormSpecMenu.cpp

```diff
@@ -2707,6 +2900,68 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
 		setInitialFocus();
 
 	skin->setFont(old_font);
+
+	// legacy sorting
+	if (m_formspec_version < 2)
+		legacySortElements(legacy_sort_start);
+}
+
+void GUIFormSpecMenu::legacySortElements(core::list<IGUIElement *>::Iterator from)
+{
+	/*
+	 * The order was:
+	 * (0. background colors and backgrounds)
+	 * 1. boxes
+	 * 2. all normal elements (like buttons)
+	 * 3. images
+	 * 4. item images
+	 * 5. item lists (TODO)
+	 */
+
+	if (from == Children.end())
+		from = Children.begin();
+	else
+		from++;
+
+	core::list<IGUIElement *>::Iterator to = Children.end();
+
+	// step 1: move all boxes to back and all images to front
+	// (drawn in back = drawn first = in front of list)
+	for (auto i = from; i != to;) {
+		if (getTypeByID((*i)->getID()) == f_Box) {
+			auto e = *i;
+			i = Children.erase(i);
```

Comment:
`std::sort` could do this. You'd only need a function to get the priority, and one to compare one element's priority with another.

Example:
https://github.com/minetest/minetest/pull/8603/files#diff-6ba6d577c564534741b34ee55c76e5b8R343
https://github.com/minetest/minetest/pull/8603/files#diff-6ba6d577c564534741b34ee55c76e5b8R446

Just for your information. I think this code will work good as well.

Reply:
Would that ensure that the order otherwise stays the same? Eg. `button[1]box[1]image[1]button[2]box[2]button[3]image[2]` has to be `box[1]box[2]button[1]button[2]button[3]image[1]image[2]` and the internal order of boxes, buttons or images here has to be 123.

Reply:
Same elements have the same priority, hence their relative positions never change.